File: frontend/src/lib/gridSquare.ts
/** Convert a 4- or 6-char Maidenhead grid to the center lat/lon of that square. */
export function gridToLatLon(grid: string): { lat: number; lon: number } | null {
  const g = grid.trim();
  if (!/^[A-Ra-r]{2}\d{2}([A-Xa-x]{2})?$/i.test(g)) return null;

  const upper = g.toUpperCase();
  let lon = (upper.charCodeAt(0) - 65) * 20 - 180;
  let lat = (upper.charCodeAt(1) - 65) * 10 - 90;
  lon += parseInt(upper[2]) * 2;
  lat += parseInt(upper[3]) * 1;

  if (g.length >= 6) {
    const subLon = upper.charCodeAt(4) - 65;
    const subLat = upper.charCodeAt(5) - 65;
    lon += subLon * (2 / 24);
    lat += subLat * (1 / 24);
    // center of subsquare
    lon += 1 / 24;
    lat += 1 / 48;
  } else {
    // center of square
    lon += 1;
    lat += 0.5;
  }

  return { lat: parseFloat(lat.toFixed(4)), lon: parseFloat(lon.toFixed(4)) };
}

export function latLonToGrid(lat: number, lon: number): string {
  const adjLon = lon + 180;
  const adjLat = lat + 90;

  const fieldLon = Math.floor(adjLon / 20);
  const fieldLat = Math.floor(adjLat / 10);
  const squareLon = Math.floor((adjLon % 20) / 2);
  const squareLat = Math.floor((adjLat % 10) / 1);
  const subLon = Math.floor(((adjLon % 20) % 2) / (2 / 24));
  const subLat = Math.floor(((adjLat % 10) % 1) / (1 / 24));

  return (
    String.fromCharCode(65 + fieldLon) +
    String.fromCharCode(65 + fieldLat) +
    squareLon.toString() +
    squareLat.toString() +
    String.fromCharCode(97 + subLon) +
    String.fromCharCode(97 + subLat)
  );
}

File: backend/app/services/grid_square.py
from __future__ import annotations


def latlon_to_grid(lat: float, lon: float) -> str:
    """Convert latitude/longitude to Maidenhead grid square (6 characters)."""
    lon = lon + 180
    lat = lat + 90

    field_lon = int(lon / 20)
    field_lat = int(lat / 10)
    square_lon = int((lon % 20) / 2)
    square_lat = int((lat % 10) / 1)
    sub_lon = int(((lon % 20) % 2) / (2 / 24))
    sub_lat = int(((lat % 10) % 1) / (1 / 24))

    grid = (
        chr(ord("A") + field_lon)
        + chr(ord("A") + field_lat)
        + str(square_lon)
        + str(square_lat)
        + chr(ord("a") + sub_lon)
        + chr(ord("a") + sub_lat)
    )
    return grid


def grid_to_latlon(grid: str) -> tuple[float, float]:
    """Convert Maidenhead grid square to latitude/longitude (center of square)."""
    grid = grid.strip().upper()
    lon = (ord(grid[0]) - ord("A")) * 20 - 180
    lat = (ord(grid[1]) - ord("A")) * 10 - 90

    if len(grid) >= 4:
        lon += int(grid[2]) * 2
        lat += int(grid[3]) * 1

    if len(grid) >= 6:
        lon += (ord(grid[4]) - ord("A")) * (2 / 24)
        lat += (ord(grid[5]) - ord("A")) * (1 / 24)
        # Center of subsquare
        lon += 1 / 24
        lat += 0.5 / 24
    elif len(grid) >= 4:
        # Center of square
        lon += 1
        lat += 0.5
    else:
        # Center of field
        lon += 10
        lat += 5

    return lat, lon
